// magpie-base/src/components/index.js
export { default as Experiment } from './Experiment';
export { default as Screen } from './Screen';
export { default as Slide } from './Slide';

export { default as CompletionInput } from './inputs/CompletionInput';
export { default as DropdownInput } from './inputs/DropdownInput';
export { default as ForcedChoiceInput } from './inputs/ForcedChoiceInput';
export { default as ImageSelectionInput } from './inputs/ImageSelectionInput';
export { default as KeypressInput } from './inputs/KeypressInput';
export { default as RatingInput } from './inputs/RatingInput';
export { default as SliderInput } from './inputs/SliderInput';
export { default as TextareaInput } from './inputs/TextareaInput';
export { default as MultipleChoiceInput } from './inputs/MultipleChoiceInput';
export { default as MultipleChoiceMatrixInput } from './inputs/MultipleChoiceMatrixInput';
export { default as SliderRangesInput } from './inputs/SliderRangesInput';
export { default as RankOrderInput } from './inputs/RankOrderInput';
export { default as MazeInput } from './inputs/MazeInput';
export { default as SerialInput } from './inputs/SerialInput';
export { default as SelfPacedReadingInput } from './inputs/SelfPacedReadingInput';

export { default as Rsvp } from './stimuli/Rsvp';
export { default as FixationCross } from './stimuli/FixationCross';
export const Synth = () => {
  return import(/* webpackPreload: true */ './stimuli/Synth');
};

export { default as Chat } from './interactive/Chat';
export { default as WaitForParticipants } from './interactive/WaitForParticipants';

export { default as LifecycleScreen } from './screens/LifecycleScreen';
export { default as ConnectInteractiveScreen } from './screens/ConnectInteractiveScreen';
export { default as AwaitIteratedResultScreen } from './screens/AwaitIteratedResultScreen';
export { default as DebugResultsScreen } from './screens/DebugResultsScreen';
export { default as SubmitResultsScreen } from './screens/SubmitResultsScreen';
export { default as PostTestScreen } from './screens/PostTestScreen';
export { default as InstructionScreen } from './screens/InstructionScreen';
export { default as EyetrackingCalibrationScreen } from './screens/EyetrackingCalibrationScreen';
export { default as EyetrackingValidationScreen } from './screens/EyetrackingValidationScreen';

export { default as ForcedChoiceScreen } from './trial_screens/ForcedChoiceScreen';
export { default as ImageSelectionScreen } from './trial_screens/ImageSelectionScreen';
export { default as TextareaScreen } from './trial_screens/TextareaScreen';
export { default as SliderScreen } from './trial_screens/SliderScreen';
export { default as SliderRangesScreen } from './trial_screens/SliderRangesScreen';
export { default as RatingScreen } from './trial_screens/RatingScreen';
export { default as CompletionScreen } from './trial_screens/CompletionScreen';
export { default as KeypressScreen } from './trial_screens/KeypressScreen';
export { default as ForcedChoiceMousetrackingScreen } from './trial_screens/ForcedChoiceMousetrackingScreen';
export { default as MultipleChoiceScreen } from './trial_screens/MultipleChoiceScreen';
export { default as MultipleChoiceMatrixScreen } from './trial_screens/MultipleChoiceMatrixScreen';
export { default as MazeScreen } from './trial_screens/MazeScreen';
export { default as SelfPacedReadingScreen } from './trial_screens/SelfPacedReadingScreen';

export { default as Wait } from './helpers/Wait';
export { default as TimerStop } from './helpers/TimerStop';
export { default as TimerStart } from './helpers/TimerStart';
export { default as Record } from './helpers/Record';
export { default as ResponseTimeStart } from './helpers/ResponseTimeStart';
export { default as MousetrackingStart } from './helpers/MousetrackingStart';
export { default as EyetrackingStart } from './helpers/EyetrackingStart';
export { default as FullscreenStart } from './helpers/FullscreenStart';
export { default as HideCursor } from './helpers/HideCursor';
export { default as PreventCopyPaste } from './helpers/PreventCopyPaste';
